wire-ios-request-strategy/Sources/Message Sending/ProteusMessagePayloadBuilder.swift

```diff
@@ -0,0 +1,169 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireProtos
+
+enum MessageEncryptorError: Error {
+    case missingValidSelfClient
+    case missingSelfDomain
+    case unableToEncryptForExternalData
+}
+
+private extension String {
+    var hexRemoteIdentifier: UInt64 {
+        let pointer = UnsafeMutablePointer<UInt64>.allocate(capacity: 1)
+        defer { pointer.deallocate() }
+        Scanner(string: self).scanHexInt64(pointer)
+        return UInt64(pointer.pointee)
+    }
+}
+
+/// Provide the payload for a given proteus message
+struct ProteusMessagePayloadBuilder {
+    var context: NSManagedObjectContext
+    var proteusService: ProteusServiceInterface
+
+    var useQualifiedIds: Bool = false
+
+    func encryptForTransport(with messageInfo: MessageInfo, externalData: Data? = nil) async throws -> Data {
+
+        // 1) encrypt the data with proteusService
+        let plainText = try messageInfo.genericMessage.serializedData()
+        let allSessionIds = messageInfo.allSessionIds()
+        let encryptedDatas = try await proteusService.encryptBatched(data: plainText, forSessions: allSessionIds)
+
+        // 2) Wrap the encryptedData in protobuf object that will be serialized
+        var messageData: Data
+        if useQualifiedIds {
+            messageData = try await qualifiedData(messageInfo: messageInfo, encryptedDatas: encryptedDatas, externalData: externalData)
+        } else {
+            messageData = try await unQualifiedData(messageInfo: messageInfo, encryptedDatas: encryptedDatas, externalData: externalData)
+        }
+
+        // Message too big?
+        if  UInt(messageData.count) > ZMClientMessage.byteSizeExternalThreshold && externalData == nil {
+            // The payload is too big, we therefore rollback the session since we won't use the message we just encrypted.
+            // This will prevent us advancing sender chain multiple time before sending a message, and reduce the risk of TooDistantFuture.
+            messageData = try await encryptForTransportExternalDataBlob(message: messageInfo.genericMessage, messageInfo: messageInfo)
+        }
+
+        // TODO: Reset all failed sessions. -> [F] why do we reset the failedToEstablishSession on all clients ??
```

Comment:
double check this

Reply:
> added a function `resetAllUserClientsFailedSessions` called before the message is sent

I don't think you should do this after each message sent, as John explained a client will have `failedToEstablishSession = true` when we get a broken prekey when trying to establish a session with the client. Resetting this flag will not repair the session.

It only makes sense to reset this flag after you've successfully established session with the client. For example after tapping "reset session" and this time you get a valid prekey.

Reply:
ok removed the code